Address issue #4504 correct incorrect error variable returned

keif888 opened the pull request:
This pull request fixes the issue identified in #4504.

It changes the returned error variable from err to result.Error when result.Error is not nil.

This allows the database error to be propagated up the call stack until the callers can handle the error appropriately with either retires or aborts.

An example of where it can be propagated up to is:
https://github.com/photoprism/photoprism/blob/98b0dc5e07ac80fe83b439478fe767f4760bc01a/internal/api/photos.go#L102

This will result in an error message back to a consumer of the api if there is an underlying database error.
Without this fix, this code is silently failing, and returning a success to the consumer if the api has failed in a database update.


I have written an automated unit test, but I have NOT included it in the pull.
That is because the automated unit test breaks an unpredictable number of other tests.
When this test is included, any other test that is running when lock_photos_for_test is holding locks will fail if it tries to access the photos table.  This is because sqllite returns "no such table: photos" for any access to the locked table, instead of waiting for the lock to be freed (or timing out) as MariaDB does.

The unit test code is as follows:
```
package entity

import (
	"testing"
	"time"

	"github.com/jinzhu/gorm"
	"github.com/stretchr/testify/assert"
)

// Function to hold a lock on the photos table for 15 seconds.  Hopefully long enough.
func lock_photos_for_test(t *testing.T, m interface{}, keyNames ...string) {

	// Extract interface slice with all values including zero.
	values, keys, err := ModelValues(m, keyNames...)

	// Has keys and values?
	if err != nil {
		t.Logf("lock_photos_for_test ModelValues Error = %s", err.Error())
		return
	} else if len(keys) != len(keyNames) {
		t.Logf("lock_photos_for_test keys issue")
		return
	}

	db_test := UnscopedDb()
	db_test.Transaction(func(tx *gorm.DB) error {
		if tx.Error != nil {
			t.Logf("lock_photos_for_test Begin = %s", tx.Error.Error())
			return tx.Error
		}

		result := tx.Model(m).Updates(values)

		if result.Error != nil {
			t.Logf("lock_photos_for_test Model Updates Error = %s", result.Error.Error())
			return result.Error
		}

		time.Sleep(15 * time.Second)
		return nil
	})

	t.Logf("lock_photos_for_test Rollback Done")
}

func TestEntityLock(t *testing.T) {
	t.Run("LockedDB", func(t *testing.T) {
		m := PhotoFixtures.Pointer("Photo01")
		updatedAt := m.UpdatedAt

		// Should be updated without any issues.
		if err := Update(m, "ID", "PhotoUID"); err != nil {
			assert.Greater(t, m.UpdatedAt.UTC(), updatedAt.UTC())
			t.Fatal(err)
			return
		} else {
			assert.Greater(t, m.UpdatedAt.UTC(), updatedAt.UTC())
			t.Logf("(1) UpdatedAt: %s -> %s", updatedAt.UTC(), m.UpdatedAt.UTC())
			t.Logf("(1) Successfully updated values")
		}

		go lock_photos_for_test(t, m, "ID", "PhotoUID")
		// Wait a bit for the other thread to start waiting.
		time.Sleep(time.Second * 2)

		// Should return an error with lock timeout.
		if err := Update(m, "ID", "PhotoUID"); err != nil {
			assert.Greater(t, m.UpdatedAt.UTC(), updatedAt.UTC())
			assert.ErrorContains(t, err, "no such table: photos") // Sql Lite doesn't have wait locking it just throws a missing table message.
			t.Logf("(2) Error was %s", err.Error())
			return
		} else {
			t.Logf("(2) Error not found")
			t.Fail()
		}
	})

}
```

Example failed tests when including this test:
```
     | time="2024-09-06T07:47:20Z" level=debug msg="counts: updated 32 labels [296.324µs]"
FAIL | Update (0.00s)
FAIL |   Update/NotUpdated (0.00s)
     |         entity_update_test.go:54: 
     |             	Error Trace:	/go/src/github.com/photoprism/photoprism/internal/entity/entity_update_test.go:54
     |             	Error:      	Error "no such table: photos" does not contain "record not found"
     |             	Test:       	TestUpdate/NotUpdated
FAIL |   Update/Photo01 (0.00s)
     |         entity_update_test.go:64: no such table: photos
FAIL |   Update/NonExistentKeys (0.00s)
     |         entity_update_test.go:92: 
     |             	Error Trace:	/go/src/github.com/photoprism/photoprism/internal/entity/entity_update_test.go:92
     |             	Error:      	Error "no such table: photos" does not contain "record not found"
     |             	Test:       	TestUpdate/NonExistentKeys
     | time="2024-09-06T07:47:22Z" level=error msg="face: failed setting embeddings (invalid embedding)"
     | time="2024-09-06T07:47:22Z" level=error msg="face: failed setting embeddings (invalid embedding)"
     | time="2024-09-06T07:47:22Z" level=error msg="face: failed setting embeddings (invalid embedding)"
FAIL | Face_ResolveCollision (0.00s)
FAIL |   Face_ResolveCollision/collision (0.00s)
     |         face_test.go:66: no such table: faces
     | time="2024-09-06T07:47:22Z" level=debug msg="faces: found no matching markers for conflict resolution (no such table: markers)"
FAIL | Face_ReviseMatches (0.00s)
     |     face_test.go:127: no such table: markers
     | time="2024-09-06T07:47:22Z" level=error msg="face: failed setting embeddings (invalid embedding)"
     | time="2024-09-06T07:47:22Z" level=error msg="face: failed setting embeddings (invalid embedding)"
     | time="2024-09-06T07:47:22Z" level=error msg="failed parsing face embedding json: json: cannot unmarshal bool into Go value of type float64"
FAIL | Face_UpdateMatchTime (0.00s)
     |     face_test.go:207: no such table: faces
     | time="2024-09-06T07:47:22Z" level=error msg="face: failed setting embeddings (embedding has invalid number of values)"
FAIL | Face_Save (0.00s)
FAIL |   Face_Save/Ok (0.00s)
     |         face_test.go:219: no such table: faces
FAIL | Face_Update (0.00s)
     |     face_test.go:238: FaceID: 3647ZAABMPFJVYSVHXY32CLNNLIZVD44
     |     face_test.go:243: no such table: faces
FAIL | Face_RefreshPhotos (0.00s)
     |     face_test.go:263: no such table: photos
     | time="2024-09-06T07:47:22Z" level=error msg="faces: failed to add MLCW5YTJDPTUH6CJRUPQTHIQYEWXDEIN (no such table: faces)"
FAIL | FirstOrCreateFace (0.00s)
FAIL |   FirstOrCreateFace/create_new_face (0.00s)
     | panic: runtime error: invalid memory address or nil pointer dereference [recovered]
     | 	panic: runtime error: invalid memory address or nil pointer dereference
     | [signal SIGSEGV: segmentation violation code=0x1 addr=0x30 pc=0xd368aa]
     | goroutine 2293 [running]:
     | testing.tRunner.func1.2({0xf2ce40, 0x265ed40})
     | 	/usr/local/go/src/testing/testing.go:1631 +0x24a
     | testing.tRunner.func1()
     | 	/usr/local/go/src/testing/testing.go:1634 +0x377
     | panic({0xf2ce40?, 0x265ed40?})
     | 	/usr/local/go/src/runtime/panic.go:770 +0x132
     | github.com/photoprism/photoprism/internal/entity.TestFirstOrCreateFace.func1(0xc002039520)
     | 	/go/src/github.com/photoprism/photoprism/internal/entity/face_test.go:271 +0x4a
     | testing.tRunner(0xc002039520, 0x1733390)
     | 	/usr/local/go/src/testing/testing.go:1689 +0xfb
     | created by testing.(*T).Run in goroutine 2292
     | 	/usr/local/go/src/testing/testing.go:1742 +0x390
     | exit status 2
FAIL | 	github.com/photoprism/photoprism/internal/entity	10.836s

```


Acceptance Criteria:

- [X] Features and enhancements must be fully implemented so that they can be released at any time without additional work
- [Not Possible] Automated unit and/or acceptance tests are mandatory to ensure the changes work as expected and to reduce repetitive manual work
- [X] Frontend components must be responsive to work and look properly on phones, tablets, and desktop computers; you must have tested them on all major browsers and different devices
- [X] Documentation and translation updates should be provided if needed
- [X] In case you submit database-related changes, they must be tested and compatible with SQLite 3 and MariaDB 10.5.12+

keif888 committed:
Issue #4505 correct incorrect error variable returned

change from err to result.Error so that the underlying error is returned to the caller.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/photoprism/photoprism?pullRequest=4505) <br/>All committers have signed the CLA.

keif888 commented:
> [![CLA assistant check](https://camo.githubusercontent.com/bb14f9741a98be03eb447ca21a4a5e96201df401e5be70f9ef52130f4e1ff903/68747470733a2f2f636c612d617373697374616e742e696f2f70756c6c2f62616467652f6e6f745f7369676e6564)](https://cla-assistant.io/photoprism/photoprism?pullRequest=4505) Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/photoprism/photoprism?pullRequest=4505) before we can accept your contribution.You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/photoprism/photoprism?pullRequest=4505) it.

![image](https://github.com/user-attachments/assets/19d7f1eb-da92-422e-bfb2-505e34384e68)

lastzero commented:
@keif888 Thank you for your detailed report and fix! I have updated the `entity.Update()` function based on this and would appreciate it if you could test these changes, see https://github.com/photoprism/photoprism/commit/be00bcf0b34d037d77360efc3e4a2c924fea1ebf.

keif888 commented:
@lastzero My updated unit test (run on it's own) passes.  I haven't found a method, other than running all the tests in series, to allow this test to be included in the main suite.

It checks that you can update the entity, then starts another thread (lock_photos_for_test) which starts a transaction, updates and then sleeps.  Then (whilst that thread is sleeping) it attempts to do an update, which fails successfully.  It then sleeps until after the rollback has happened, and re-reads the original photo to confirm that it has reverted to the 1st update state.

```
START| EntityLock
START|   EntityLock/LockedDB
     |     entity_update_lock_test.go:65: (1) UpdatedAt: 2020-03-28 14:06:00 +0000 UTC -> 2024-09-12 13:00:28.885121425 +0000 UTC
     |     entity_update_lock_test.go:66: (1) Successfully updated values
     |     entity_update_lock_test.go:14: lock_photos_for_test Starting
     |     entity_update_lock_test.go:34: lock_photos_for_test Before updating the photo
     |     entity_update_lock_test.go:36: lock_photos_for_test After updating the photo
     |     entity_update_lock_test.go:43: lock_photos_for_test Sleeping
     |     entity_update_lock_test.go:77: (2) UpdatedAt: 2024-09-12 13:00:28.885121425 +0000 UTC -> 2024-09-12 13:00:29.891874187 +0000 UTC
     |     entity_update_lock_test.go:78: (2) Error was no such table: photos
     |     entity_update_lock_test.go:45: lock_photos_for_test Awake, rolling back now
     |     entity_update_lock_test.go:50: lock_photos_for_test Rollback Done
     |     entity_update_lock_test.go:83: (3) UpdatedAt: 2024-09-12 13:00:28.885121425 +0000 UTC -> 2024-09-12 13:00:28.885121425 +0000 UTC
PASS | EntityLock (9.01s)
PASS |   EntityLock/LockedDB (9.01s)
PASS | github.com/photoprism/photoprism/internal/entity 
```


My manual test, which was to lock the label record for sunglasses in the mariadb backend, and update the location of a photo with sunglasses passed.
```
DEBU[2024-09-12T13:11:03Z] server: GET /api/v1/photos/psjccwm9zyi80u08 (200) [11.352989ms] 
DEBU[2024-09-12T13:11:16Z] place: us:auQ0qOgDgMAq not found             
INFO[2024-09-12T13:11:16Z] cell: added place us:auQ0qOgDgMAq [1.20837364s] 
DEBU[2024-09-12T13:11:16Z] cell: added s2:89c624edf474 [1.210217159s]   
DEBU[2024-09-12T13:11:16Z] photo: changing location of 2024/09/20240905_131637_D0BCE4EF from s2:89c896861554 to s2:89c624edf474 
DEBU[2024-09-12T13:11:16Z] photo: 2024/09/20240905_131637_D0BCE4EF title based on label Sports 
DEBU[2024-09-12T13:11:16Z] photo: 2024/09/20240905_131637_D0BCE4EF has new title 'Sports / Lancaster / 2024' [848.277µs] 
DEBU[2024-09-12T13:17:16Z] api-v1: abort /api/v1/photos/:uid with code 500 (changes could not be saved) 
DEBU[2024-09-12T13:17:16Z] server: PUT /api/v1/photos/psjccwm9zyi80u08 (500) [6m1.303514024s] 
```
Note the api-v1: abort message 2nd from last above, which shows that the error message was propagated up successfully.

keif888 commented:
Updated unit test code:
```
package entity

import (
	"testing"
	"time"

	"github.com/jinzhu/gorm"
	"github.com/stretchr/testify/assert"
)

// Function to hold a lock on the photos table for 15 seconds.  Hopefully long enough.
func lock_photos_for_test(t *testing.T, m interface{}, keyNames ...string) {

	t.Logf("lock_photos_for_test Starting")
	// Extract interface slice with all values including zero.
	values, keys, err := ModelValues(m, keyNames...)

	// Has keys and values?
	if err != nil {
		t.Logf("lock_photos_for_test ModelValues Error = %s", err.Error())
		return
	} else if len(keys) != len(keyNames) {
		t.Logf("lock_photos_for_test keys issue")
		return
	}

	db_test := UnscopedDb()
	db_test.Transaction(func(tx *gorm.DB) error {
		if tx.Error != nil {
			t.Logf("lock_photos_for_test Begin = %s", tx.Error.Error())
			return tx.Error
		}

		t.Logf("lock_photos_for_test Before updating the photo")
		result := tx.Model(m).Updates(values)
		t.Logf("lock_photos_for_test After updating the photo")

		if result.Error != nil {
			t.Logf("lock_photos_for_test Model Updates Error = %s", result.Error.Error())
			return result.Error
		}
		
		t.Logf("lock_photos_for_test Sleeping")
		time.Sleep(4 * time.Second)
		t.Logf("lock_photos_for_test Awake, rolling back now")
		tx.Rollback()
		return nil
	})

	t.Logf("lock_photos_for_test Rollback Done")
}

func TestEntityLock(t *testing.T) {
	t.Run("LockedDB", func(t *testing.T) {
		m := PhotoFixtures.Pointer("Photo01")
		updatedAt := m.UpdatedAt

		// Should be updated without any issues.
		if err := Update(m, "ID", "PhotoUID"); err != nil {
			assert.Greater(t, m.UpdatedAt.UTC(), updatedAt.UTC())
			t.Fatal(err)
			return
		} else {
			assert.Greater(t, m.UpdatedAt.UTC(), updatedAt.UTC())
			t.Logf("(1) UpdatedAt: %s -> %s", updatedAt.UTC(), m.UpdatedAt.UTC())
			t.Logf("(1) Successfully updated values")
		}
		updatedAt = m.UpdatedAt
		go lock_photos_for_test(t, m, "ID", "PhotoUID")
		// Wait a bit for the other thread to start waiting.
		time.Sleep(time.Second * 1)

		// Should return an error with lock timeout.
		if err := Update(m, "ID", "PhotoUID"); err != nil {
			assert.Greater(t, m.UpdatedAt.UTC(), updatedAt.UTC())
			assert.ErrorContains(t, err, "no such table: photos") // Sql Lite doesn't have wait locking it just throws a missing table message.
			t.Logf("(2) UpdatedAt: %s -> %s", updatedAt.UTC(), m.UpdatedAt.UTC())
			t.Logf("(2) Error was %s", err.Error())
			// Wait for the rollback to be done.
			time.Sleep(time.Second * 8)
			// Find the record that was there before the lock
			n := m.Find()
			t.Logf("(3) UpdatedAt: %s -> %s", updatedAt.UTC(), n.UpdatedAt.UTC())
			assert.Equal(t, n.UpdatedAt.UTC(), updatedAt.UTC())
			return
		} else {
			t.Logf("(2) Error not found")
			t.Fail()
		}
	})

}
```

keif888 committed:
Merge remote-tracking branch 'upstream/develop' into Issue-#4504

lastzero commented:
Hi @keif888! This PR has the label *released*, but there seem to be new changes? If these are also included in #4831, I would like to close this PR to avoid confusion... :thinking:

keif888 commented:
Hi Michael,
This has been released, and I can see the changes in the Develop branch.
As it's in the develop branch, it is in the GORMv2 and PostgreSQL PR's as well.
I have also included unit tests for the scenario.

Closed.
